File: 34/fifth_powers.py
#!/usr/bin/python

def factorial(n):
  if n == 0:
    return 1
  else:
    return n * factorial(n - 1)

def special(n):
  return n == sum([factorials[int(x)] for x in str(n)])

factorials = [factorial(n) for n in range(10)]
result = []
upper_bound = 7 * factorials[9]

for i in range(3, upper_bound):
  if special(i):
    result.append(i)

print("There are {} values satisfying this condition: {}.".format(len(result), ",".join(list(map(str, result)))))
print("Their sum is {}.".format(sum(result)))

File: 34/fifth_powers.rb
#!/usr/bin/ruby

def factorial(n)
  return 1 if n == 0
  n * factorial(n - 1)
end

def special(n)
  n == n.to_s.split('').map{|ch| factorial(ch.to_i)}.reduce(:+)
end

factorials = (0..9).to_a.map{|i| factorial(i)}
result = []
upper_bound = 7 * factorials[9]

for i in (3..upper_bound).to_a do
  result.push(i) if special(i)
end

puts("There are #{result.length} values satisfying this condition: #{result}.")
puts("Their sum is #{result.reduce(:+)}.")
